cranelift/codegen/src/machinst/abi_impl.rs

```diff
@@ -1439,14 +1556,55 @@ impl<M: ABIMachineSpec> ABICaller for ABICallerImpl<M> {
                     // Store the extended version.
                     ty = M::word_type();
                 }
-                for insn in gen_store_stack_multi::<M>(StackAMode::SPOffset(off, ty), from_regs, ty)
+                for insn in
+                    gen_store_stack_multi::<M>(StackAMode::SPOffset(offset, ty), from_regs, ty)
+                {
+                    ctx.emit(insn);
+                }
+            }
+            &ABIArg::StructArg { offset, size, .. } => {
+                let src_ptr = from_regs.only_reg().unwrap();
+                let dst_ptr = ctx.alloc_tmp(M::word_type()).only_reg().unwrap();
+                ctx.emit(M::gen_get_stack_addr(
+                    StackAMode::SPOffset(offset, I8),
+                    dst_ptr,
+                    I8,
+                ));
+                // Emit a memcpy from `src_ptr` to `dst_ptr` of `size` bytes.
+                // N.B.: because we process StructArg params *first*, this is
+                // safe w.r.t. clobbers: we have not yet filled in any other
+                // arg regs.
+                //
+                // TODO: we assume that the signature of the callee is also the
+                // signature of the memcpy() function. Can we do better?
```

Comment:
Just saw this TODO question: for the calling convention, there's actually `CallConv::for_libcall`. For the rest of the signature, we should assert in the `gen_memcpy` impl that the calling convention is one of a few known (e.g., not any of the `Baldrdash` ones).

Reply:
Ah, that's more canonical, yes. Updated this and added asserts in x64/aarch64 `gen_memcpy()` that we aren't using a Baldrdash calling convention -- we should never be seeing struct args there anyway.